### PR TITLE
feat(authorization): add uri validation on the authorization endpoint

### LINKED ISSUE #387: Update Joi version to support uri validation
The new version of Joi supports [uri validation](https://github.com/hapijs/joi/blob/v7.2.3/API.md#stringurioptions). Could use this for `redirectUri` validation

https://github.com/mozilla/fxa-oauth-server/blob/master/lib/routes/authorization.js#L104

### LINKED ISSUE #388: restrict redirect_uri to `https://` and `urn:`
This extends from [this comment in the content server](https://github.com/mozilla/fxa-content-server/pull/3469#discussion_r52017618). @vladikoff mentioned the OAuth server allows any arbitrary string as the redirect_uri in the call to `/authorization`, which is exceptionally lenient. @rfk and I talked today about tightening this up to only allow `https://` and `urn:` URIs, and additional protocols can be added if required.

I imagine both the `redirect_uri` passed to `/authorization` and the `redirect_uri` registered for each client would both need to be validated.

Ref https://github.com/mozilla/fxa-content-server/pull/3469, #387

### PR DESCRIPTION
**Pull request by vladikoff**
Fixes #387
Fixes #388 

@vbudhram r?

### CODE REVIEW

**Comment by seanmonstar** [comment]
This doesn't cause problems with `urn:ietf:wg:oauth:2.0:fx:webchannel` that Firefox uses?

**Comment by vladikoff** [comment]
> This doesn't cause problems with urn:ietf:wg:oauth:2.0:fx:webchannel that Firefox uses?

good question, I think only Hello used it and we should not have anymore OAuth reliers that use that. 

Ref https://github.com/mozilla/fxa-content-server/commit/c07664f#diff-f243a8eaa90957e9b805f13009c69169

**Comment by vladikoff** [comment]
This PR also addresses https://github.com/mozilla/fxa-oauth-server/issues/388 somewhat. 

Should we allow `http` here?

**Comment by vladikoff** [comment]
@rfk  ^ ?

**Comment by vbudhram** [comment]
@seanmonstar Yea, there used be to validation for URN redirects previously but it was removed when support for oauth webchannel keys was dropped, https://github.com/mozilla/fxa-content-server/pull/4295

**Comment by rfk** [comment]
> This doesn't cause problems with urn:ietf:wg:oauth:2.0:fx:webchannel that Firefox uses?

Just to confirm, yes, we deliberately removed this as we don't have any webchannel oauth reliers anymore.  We can bring it back later if we need it.

> Should we allow http here? Or should http be only for latest and stable, but not stage and prod? 

Any chance we can restrict the http URLs to localhost only?  Otherwise, I'd be happy with the ability to pref off support for http URLs in production config.  Production reliers should absolutely not be using http redirect URIs (with the possible exception of localhost).

File: lib/routes/authorization.js [hunk 1]
```diff
@@ -102,7 +102,14 @@ module.exports = {
       assertion: validators.assertion
         .required(),
       redirect_uri: Joi.string()
-        .max(256),
+      .max(256)
```
Comment (seanmonstar) [inline lib/routes/authorization.js]: I agree that consistency in style should be kept.

**Comment by vladikoff** [comment]
@rfk and @vladikoff to discuss this...

**Comment by rfk** [comment]
Whoops, forgot to discuss this in 1:1.  It sounds like the simplest option will be to allow http:// URLs on dev boxes, but only https:// in stage on prod.  @vladikoff thoughts?

**Comment by vladikoff** [comment]
> Whoops, forgot to discuss this in 1:1. It sounds like the simplest option will be to allow http:// URLs on dev boxes, but only https:// in stage on prod. @vladikoff thoughts?

Yes, that's a good plan

**Comment by vladikoff** [comment]
@seanmonstar updated!

**Comment by vladikoff** [comment]
Filed https://github.com/mozilla/fxa-dev/pull/288 to allow `http` clients in fxa-dev

File: config/dev.json [hunk 2]
```diff
@@ -95,5 +95,6 @@
   "openid": {
     "keyFile": "../config/key.json",
     "key": {}
-  }
+  },
+  "fxaEnv": "development"
```
Comment (vladikoff) [inline config/dev.json]: `env` is node environment, fxaEnv is Firefox Accounts running mode. 

We discussed adding this in the future to stop relying on node's env. @jrgm thoughts?

Comment (vladikoff) [inline config/dev.json]: from meeting: let's make a new env variable for this feature and enable it in fxa-dev.

**Pull request merged**